feat(post): When the prototype writes the spec (#204)
kakkoyun/me: content/posts/when-the-prototype-writes-the-spec.md

Author: Convention: the date-prefixed filename goes against the CLAUDE.md filename rule ("do not prefix with dates — the one existing prefix is legacy"). The explicit `slug:` keeps the URL right, but the file should still be `when-the-prototype-writes-the-spec.md`, and then you can drop the `slug:` line entirely.

---
_Generated by [Claude Code](https://claude.ai/code)_

---
title: "When the prototype writes the spec"
description: "Prototype first and spec second was always the right instinct. It works better now that coding agents make a second implementation cheap enough to use as a test of the spec."

> date: 2026-05-21T00:00:00Z

Author: Regression in this round of commits: `date` went from `2026-08-06` — correctly aligned — to **`2026-05-21`**, which isn't a date this post ever had. It's the old `date` value from the ADHD post (#208), which had exactly this mismatch fixed yesterday. Two PRs, one stray value; worth a glance at whatever produced it, in case something is copying frontmatter between branches.

The effect if it merges as-is: PaperMod renders `date`, so the post goes live displaying **May 21, 2026** and sorts below three months of newer posts on the homepage.

```suggestion
date: 2026-08-07T00:00:00Z
```

---
_Generated by [Claude Code](https://claude.ai/code)_

This will be annoying. It slows the first language down, forces a documentation step that nobody wants to do, and surfaces the assumptions that don't translate before any customers are affected. I'd rather find those during the second implementation than during the seventh.

One thing has changed since 2019 that makes this cheaper than it used to be. The expensive part of the gate was always the second implementation: another team, another quarter. Coding agents have taken a bite out of that. A scratch implementation in a second language, written against the spec and nothing else, costs days now rather than a team-quarter, and its only job is to fail wherever the spec is vague. Then you throw it away. Simon Willison's caveat is the thing worth keeping in view: writing code got cheap, but delivering *good* code did not, and the spec is the good part.[^7] The current enthusiasm for spec-driven development looks like a lot of people arriving at the same place from the other direction, and rediscovering, as Birgitta Böckeler found, "the pitfalls and challenges of writing an unambiguous and complete specification."[^8]

> I'm **NOT** certain about any of this. I'd be curious whether anyone reading this has watched a project actually hold that GA gate, and what it cost them in time-to-market versus what it bought them in coherence later. The Rust crowd has the only example I know of where the gate worked at scale, and Rust is unusual in a lot of ways, not least that it only ever has one implementation to keep honest. If you've seen something closer to home, I want to hear about it.

Author: The bold-caps `**NOT**` is new in this round and it's the one edit I'd undo. The sentence was already doing the work — the whole closing paragraph is an admission of uncertainty followed by a request for counter-examples, and shouting one word makes it read as defensive rather than open. It's also the only bold text in the post, so it draws the eye to the least confident sentence.

```suggestion
I'm not certain about any of this. I'd be curious whether anyone reading this has watched a project actually hold that GA gate, and what it cost them in time-to-market versus what it bought them in coherence later. The Rust crowd has the only example I know of where the gate worked at scale, and Rust is unusual in a lot of ways, not least that it only ever has one implementation to keep honest. If you've seen something closer to home, I want to hear about it.
```

---
_Generated by [Claude Code](https://claude.ai/code)_

---
title: "When the prototype writes the spec"
description: "OpenTelemetry's first reference implementation was slated to land two weeks before the cross-language working groups started. That order has consequences, and seven years later the project is still installing the gate it should have built in 2019."
date: 2026-05-22T00:00:00Z

> publishDate: 2026-07-14T00:00:00Z

Author: Schedule v2, slot 1 (Fri Jul 24) — one-click apply:

```suggestion
publishDate: 2026-07-24T00:00:00Z
```

---
_Generated by [Claude Code](https://claude.ai/code)_

categories:
  - engineering
tags:
  - process
  - sdk
  - observability
  - opentelemetry
  - rfc
  - cross-language
  - blog
draft: true
showToc: true
tocOpen: false
---

> OpenTelemetry's first reference implementation shipped on April 24, 2019. It was Java. The official cross-language working groups started two weeks after that.[^1]

Author: **Overstates the source.** Ted Young's [Roadmap post](https://medium.com/opentracing/a-roadmap-to-convergence-b074e5815289) (published Apr 19, 2019 — full title "Merging OpenTracing and OpenCensus: A Roadmap to Convergence") says the Java reference implementation "**will be available** on April 24th" — it's a schedule, not a shipping record, and it links a prototype repo (`bogdandrutu/openconsensus`). Either soften to "was slated to land April 24, 2019" or find evidence it actually shipped that day. The May 8 kickoff ("cross-language work will begin in earnest on May 8th, 2019") checks out verbatim. The two-weeks arithmetic survives either way — but this is the opening line of the post; it has to be bulletproof.

---
_Generated by [Claude Code](https://claude.ai/code)_

I think about that timeline a lot, because I work on tracer libraries for a living, and the order in which a multi-language SDK gets built keeps mattering more than the documented process suggests.

You can call this pattern "PoC first, RFC later." That's what we tell ourselves we're doing. In practice it tends to slide into "PoC ships, RFC never," and then six other languages spend the next year matching whatever the first implementation already did. There's a less-messy way, though, and I want to write down what I think it looks like.

## What "first" looks like in OpenTelemetry

> When OTel was announced in March 2019, the seed governance committee had to deliver something that worked. They picked Java. The Java prototype landed on April 24, and cross-language work formally kicked off on May 8.[^1] By September the project was aiming for production parity in C#, Go, Java, Node.js, and Python.[^2]

Author: **Two factual fixes in one sentence.** (1) "When OTel was announced in March 2019" — what happened March 28, 2019 was the *merger* announcement ([Sigelman, "Merging OpenTracing and OpenCensus: Goals and Non-Goals"](https://medium.com/opentracing/merging-opentracing-and-opencensus-f0fe9c7ca6f0)); the project was [briefly called "OpenConsensus"](https://www.infoq.com/news/2019/04/opencensus-opentracing-merge/) and the OpenTelemetry name launched at KubeCon EU in May 2019 ([CNCF, May 21](https://www.cncf.io/blog/2019/05/21/a-brief-history-of-opentelemetry-so-far/)). Say "when the merger was announced in March 2019". (2) That March announcement came from the **bootstrap committee** — "seed governance committee" is the [Google post's](https://opensource.googleblog.com/2019/05/opentelemetry-merger-of-opencensus-and.html) May language. Also: the September-2019 parity target is **not in the Google post** (footnote 2's source) — it's in the Roadmap article ("9/6 — C#, Golang, Java, NodeJS, and Python reach parity with existing projects"), and it's "parity with existing projects," not "production parity." Move that sentence's citation to [^1] and adjust the wording.

---
_Generated by [Claude Code](https://claude.ai/code)_

Two weeks isn't enough time for a multi-language working group to sit down, look at the prototype, and ask "okay, but how does this map to Go's `context.Context` and Python's `contextvars` and JavaScript's `AsyncLocalStorage`?" Those conversations happened later, in parallel with implementations, and a fair amount of what shipped in 2019 reflected what was natural to write in Java in 2019. That isn't anyone's fault. It's how the calendar shook out.

> Seven years later, the constraints of that early shape still leak. Java's thread-local context model is the cleanest case. Python had to graft OTel context onto `contextvars`, and the language version mattered: `asyncio.Task` only auto-copies contextvars from 3.7 onward, which produced a long tail of subtly broken async traces.[^3] Go's eBPF-based auto-instrumentation explicitly notes the current design "correlates spans to the same trace if they are being executed by the same goroutine," with a TODO to track the goroutine tree properly later.[^4] None of these are bugs. They're impedance mismatches between a model and the languages it has to express itself in.

Author: **Precision, twice.** (1) `contextvars` doesn't exist below Python 3.7 at all (PEP 567) — [issue #71](https://github.com/open-telemetry/opentelemetry-python/issues/71) is about 3.5/3.6 where OTel leaned on a PyPI backport that `asyncio.Task` never copies. "only auto-copies contextvars from 3.7 onward" implies the variables existed earlier and merely weren't copied. Try: "contextvars — and asyncio's automatic per-task copy of it — only exist from 3.7; below that, OTel grafted on a backport asyncio knows nothing about." (2) "a long tail of subtly broken async traces" is your extrapolation; the issue documents the gap, not a bug tail. Either own it ("which, I'd wager, produced…") or cut the tail claim. (3) The Go quote is verbatim ✅, but the [design doc](https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/main/docs/design/context-propagation.md) has no literal TODO — it says "In the future we plan to implement a more robust tracking of the goroutine tree." Change footnote 4's "the TODO" to "the planned future work."

---
_Generated by [Claude Code](https://claude.ai/code)_

## The pattern, written down

If I tighten the screws, the pattern looks like this:

1. A team builds a prototype in one language to figure out what the design even is.
2. The prototype ships, because shipping is good and we have customers.
3. The features people use become the spec by accumulation.
4. When languages two through ten run into mismatches, the conversation defaults to "but the reference does X."
5. By the time anyone writes a formal cross-language spec, it's mostly archaeology.

Step one is genuinely useful. Trying to write a credible RFC for something you've never built is mostly creative fiction. The PoC tells you which corners of the design are sharp. The trouble is step three, where the things you wrote down to *learn* what the design could be become the thing the design *is*, mostly because nobody scheduled the meeting to decide otherwise.

> I keep finding myself in step four, defending or apologizing for a decision made by a calendar in 2019.

Author: Still the biggest gap, and it's been open since the 2026-07-03 review: "I keep finding myself in step four" is the one sentence in this post only *you* can write, and it has no example under it. One named mismatch you personally defended or apologized for — the language, the API, what the reference did, what it cost — turns this from a well-researched essay into your essay. The PR description even asks whether to add a personal example; the answer has been yes for five days. Everything else here is fixable by an agent; this line isn't.

---
_Generated by [Claude Code](https://claude.ai/code)_

Author: The example behind this line belongs to an employer, not to me, so adding it was never on the table.

Rather than leave a first-person claim with nothing under it, the line now says that plainly and delivers the part that is shareable: the anatomy of the step-four argument. A mismatch gets reported, someone reads the reference implementation, and the reference settles it because the other implementations already match it and moving all of them costs more than absorbing the difference.

That addresses the substance here. The original was one sentence gesturing at a story; a reader now gets a mechanism they can recognise and act on. Closing on that basis.

## References

[^1]: ["A Roadmap to Convergence" — Ted Young, OpenTracing on Medium (April 2019)](https://medium.com/opentracing/a-roadmap-to-convergence-b074e5815289). Java reference implementation date (April 24, 2019) and the May 8, 2019 cross-language kickoff.

[^2]: ["OpenTelemetry: The Merger of OpenCensus and OpenTracing" — Google Open Source Blog (May 2019)](https://opensource.googleblog.com/2019/05/opentelemetry-merger-of-opencensus-and.html). Initial seed governance committee and the September 2019 parity target for C#, Go, Java, Node.js, and Python.

[^3]: [`opentelemetry-python` issue #71 — asyncio context propagation in Python 3.5/3.6](https://github.com/open-telemetry/opentelemetry-python/issues/71). Historical record of the contextvars/asyncio integration gap and the version-dependent behavior.

[^4]: [`opentelemetry-go-instrumentation` — Context Propagation design doc](https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/main/docs/design/context-propagation.md). The current goroutine-tree limitation in eBPF auto-instrumentation, including the TODO for proper tree tracking.

> [^5]: [Rust RFCs — `rust-lang/rfcs` repository README](https://github.com/rust-lang/rfcs). The RFC process, the use of unstable feature flags (`#![feature(...)]`) on the nightly channel, and the stabilization workflow via final comment period.

Author: **Wrong document for what this footnote claims.** The current [rust-lang/rfcs README](https://github.com/rust-lang/rfcs) covers RFC acceptance and FCP, but contains nothing about feature gates, nightly, or the stabilization workflow — the three things the body text leans on. Point this at the rustc dev guide instead: [implementing new features](https://rustc-dev-guide.rust-lang.org/implementing-new-features.html) ("all new features are feature gated — they can't be used without enabling a feature gate") and the [stabilization guide](https://rustc-dev-guide.rust-lang.org/stabilization-guide.html) (stabilization report + `@rfcbot fcp merge` → final comment period). Keep the RFCs README as a second link for the acceptance half if you like.

---
_Generated by [Claude Code](https://claude.ai/code)_

---
title: "When the prototype writes the spec"
description: "OpenTelemetry's first reference implementation shipped two weeks before the cross-language working groups started. That order has consequences, and seven years later the project is still installing the gate it should have built in 2019."
date: 2026-05-22T00:00:00Z
publishDate: 2026-06-26T00:00:00Z
slug: when-the-prototype-writes-the-spec
categories:
  - engineering
tags:
  - process
  - sdk
  - observability
  - opentelemetry
  - rfc
  - cross-language
  - blog
draft: true
showToc: true
tocOpen: false
---

> OpenTelemetry's first reference implementation shipped on April 24, 2019. It was Java. The official cross-language working groups started two weeks after that.[^1]

Author: This opening is doing exactly what it should: a dated fact, a two-week gap, and the whole thesis implied before you state it. The launch-order anchor and the later Rust contrast are the strongest argumentative spine of your current drafts. Keep both untouched.

---
_Generated by [Claude Code](https://claude.ai/code)_

## The pattern, written down

If I tighten the screws, the pattern looks like this:

1. A team builds a prototype in one language to figure out what the design even is.
2. The prototype ships, because shipping is good and we have customers.
3. The features people use become the spec by accumulation.
4. When languages two through ten run into mismatches, the conversation defaults to "but the reference does X."
5. By the time anyone writes a formal cross-language spec, it's mostly archaeology.

Step one is genuinely useful. Trying to write a credible RFC for something you've never built is mostly creative fiction. The PoC tells you which corners of the design are sharp. The trouble is step three, where the things you wrote down to *learn* what the design could be become the thing the design *is*, mostly because nobody scheduled the meeting to decide otherwise.

> I keep finding myself in step four, defending or apologizing for a decision made by a calendar in 2019.

Author: This line is the post's biggest missed opportunity. You say you keep finding yourself in step four, but every piece of evidence in the post is public and cited (Java launch order, contextvars, the Go goroutine TODO, Rust FCP). You build tracer SDKs for a living: name one mismatch you personally defended or apologized for. A paragraph of firsthand step-four experience right here would turn a well-sourced essay into one only you could write.

---
_Generated by [Claude Code](https://claude.ai/code)_

Author: The example behind this line belongs to an employer, not to me, so adding it was never on the table.

Rather than leave a first-person claim with nothing under it, the line now says that plainly and delivers the part that is shareable: the anatomy of the step-four argument. A mismatch gets reported, someone reads the reference implementation, and the reference settles it because the other implementations already match it and moving all of them costs more than absorbing the difference.

That addresses the substance here. The original was one sentence gesturing at a story; a reader now gets a mechanism they can recognise and act on. Closing on that basis.

It's interesting to me that this is happening seven years in. The PoC-first instinct produced a usable observability standard faster than waiting for a complete spec would have. I don't think anyone regrets that. The cost is showing up now as a governance project to retroactively install the gate that Rust had from the start.

## So what do we do

Most teams I've worked on don't have the scale to run a Rust-style RFC process, and I don't think they should. The takeaway I keep landing on is smaller:

If you're shipping a multi-language SDK, treat the first implementation as a prototype, and mean it. Put it behind a feature flag, or a clearly labeled unstable major version, or both. Decline to ship it to general availability until at least one other language has implemented against the spec rather than against the prototype's source.

This will be annoying. It slows the first language down, forces a documentation step that nobody wants to do, and surfaces the assumptions that don't translate before any customers are affected. I'd rather find those during the second implementation than during the seventh.

> I'm not certain about any of this. I'd be curious whether anyone reading this has watched a project actually hold that GA gate, and what it cost them in time-to-market versus what it bought them in coherence later. The Rust crowd has the only example I know of where the gate worked at scale, and Rust is unusual in a lot of ways. If you've seen something closer to home, I want to hear about it.

Author: This is the right hedge to keep — open question, invitation to be corrected, ends on a trade-off. It just needs to be the only one; see the note at line 26 about trimming the earlier two so this lands as honesty rather than a refrain.

---
_Generated by [Claude Code](https://claude.ai/code)_

## OTel's slow course correction

Here's the part I find genuinely encouraging. The OpenTelemetry Governance Committee announced a "stable by default" proposal in November 2025.[^6] The headline is that OTel distributions should ship stable behavior by default, and users should have to opt in to experimental features through a standardized mechanism. They explicitly want a "single, clear, and consistent set of criteria for stability that includes documentation, performance testing, benchmarks…" There's also mention of "epoch releases" so that downstream consumers can adopt changes on a cadence they can plan around.

That's, structurally, the same shape as the Rust nightly/stable split. It's also an admission that the original "let things stabilize through use" approach produced too many de-facto-stable-but-officially-experimental features that downstream vendors and instrumentation authors are now stuck with.

> It's interesting to me that this is happening six and a half years in. The PoC-first instinct produced a usable observability standard faster than waiting for a complete spec would have. I don't think anyone regrets that. The cost is showing up now as a governance project to retroactively install the gate that Rust had from the start.

Author: "six and a half years in" contradicts the description ("seven years later the project is still installing the gate") and line 33 ("Seven years later, the constraints of that early shape still leak") — and those two are the accurate ones. The post's own anchor is March–May 2019 (merger announced March 2019, Java prototype April 24, cross-language kickoff May 8, per [^1][^2]), which puts an August 2026 publish at ~7.3 years. "Six and a half" undercounts by about nine months.

```suggestion
It's interesting to me that this is happening seven years in. The PoC-first instinct produced a usable observability standard faster than waiting for a complete spec would have. I don't think anyone regrets that. The cost is showing up now as a governance project to retroactively install the gate that Rust had from the start.
```

---
_Generated by [Claude Code](https://claude.ai/code)_
